## Vectors

A vector is a quantity that has magnitude and direction.
Examples:
[3, 2]
[1, 4, 5]

Vectors are commonly used in ML to represent Data

#### Magnitude of a Vector

For a vector [x, y]

magnitude = √(x² + y²)

In [7]:
import numpy as np

vector = np.array([3, 8])
magnitude = np.linalg.norm(vector)

print(magnitude)

8.54400374531753


#### What I understood

The vector [3, 8] has a magnitude of 8.54. The magnitude of a vector, tells us the length of the vector

## Matrices

A matrix is a two dimensional array of numbers arranged in rows and columns. Each entry within a matrix is referenced by its row index and column index. 

Vectors can be understood as a special case of matrices where a column vector would be m x 1 matrix and a row vector would be a 1 x n matrix.

Examples:
-[[5,7],
  [8,10]]

In ML, matrices are commonly used to represent datasets as design matrices(where rows represent N data samples and columns represent D features), store network layer weights and map input vectores to output through linear transformations. 

#### Transpose of a matrix

The transpose of a matrix is basically a flipped format(across the main diagonal) of it, i.e., the rows become the columns and vice versa.

In [15]:
matrix = np.array([[[5,3,7],[8,10]]],dtype=object) #forcing an object array(jagged array) helps in creating a non-uniform matrix and avoids value error.

#transpose the matrix
matrix_transpose = np.transpose(matrix)

print("Original Matrix Shape:\n", matrix.shape)
print("Transposed Matrix Shape:\n", matrix_transpose.shape)

Original Matrix Shape:
 (1, 2)
Transposed Matrix Shape:
 (2, 1)


#### What I understood

Transposing the matrix flipped the structure, as the shape is reversed. The 1st row becomes the 1st column and the 2nd row become the 2nd column